### PR TITLE
Skip replacing dataloader sampler if it's already a distributed sampler

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

This PR skips replacing the dataloader's sampler if the sampler is already of type `DistributedSampler`. Currently this raises a `MisconfigurationException`. 

This causes confusion, since `trainer.replace_sampler_ddp` is by default `True`. In these cases, we can silently skip replacing the sampler instead of failing.

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by s-rog** [comment]
Has this change already been approved by one of our team members?
Currently `trainer.replace_sampler_ddp` is supposed to be `False` in this case, and using the default `True` is considered user error.

**Review by justusschock — APPROVED** [review]

**Comment by justusschock** [comment]
@s-rog while this is true, we still can do that type check and the main idea of that flag is being able to turn it off (which you still can)

**Review by SkafteNicki — APPROVED** [review]

**Review by s-rog — COMMENTED** [review]
Do we already have a test that covers this new behavior? Otherwise LGTM!

**Comment by ananthsub** [comment]
@s-rog I'll add one for a dataloader configured with a distributed sampler and `replace_sampler_ddp=True` to confirm it doesn't throw

**Comment by awaelchli** [comment]
Be careful, with this change, the docs are slightly misleading now. It currently reads:

> Enables auto adding of distributed sampler. By default it will add shuffle=True for train sampler and shuffle=False for val/test sampler. If you want to customize it, you can set replace_sampler_ddp=False and add your own distributed sampler.

File: tests/trainer/test_dataloaders.py [hunk 1]
```diff
@@ -746,6 +746,39 @@ def test_dataloader_distributed_sampler(tmpdir):
...
+            )
+
+    model = ModelWithDataLoaderDistributedSampler()
+    trainer = Trainer(
+        gpus=[0, 1],
+        num_nodes=1,
+        distributed_backend='ddp_spawn',
+        default_root_dir=tmpdir,
+        max_steps=100,
+        callbacks=[DistribSamplerCallback()]
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:774]:
```suggestion
        callbacks=[DistribSamplerCallback()],
```

**Review by s-rog — APPROVED** [review]

File: tests/trainer/test_dataloaders.py [hunk 2]
```diff
@@ -746,6 +746,39 @@ def test_dataloader_distributed_sampler(tmpdir):
     trainer.test(ckpt_path=None)
 
 
+@pytest.mark.skipif(platform.system() == 'Windows', reason='Does not apply to Windows platform.')
+@pytest.mark.skipif(torch.cuda.device_count() < 2, reason='Test requires multiple GPUs')
+def test_dataloader_distributed_sampler_already_attached(tmpdir):
+    """ Test DistributedSampler and it's arguments for DDP backend when DistSampler already included on dataloader """
+
+    class ModelWithDataLoaderDistributedSampler(EvalModelTemplate):
```
Comment (awaelchli) [inline tests/trainer/test_dataloaders.py:754]: hey @ananthsub 
Looks like you need to move this class outside the test function or we get the good old pickling error :)

**Comment by carmocca** [comment]
LGTM. Great idea!

**Review by SeanNaren — APPROVED** [review]

**Review by SeanNaren — CHANGES_REQUESTED** [review]
We need to fix the test, seems like its failing on multiple GPUs.

File: tests/trainer/test_dataloaders.py [hunk 3]
```diff
@@ -746,6 +746,39 @@ def test_dataloader_distributed_sampler(tmpdir):
...
+
+    def train_dataloader(self):
+        dataloader = super().train_dataloader()
+        dist_sampler = DistributedSampler(dataloader.dataset, shuffle=False)
+        return DataLoader(
+            dataloader.dataset,
+            batch_size=self.batch_size,
+            drop_last=False,
+            sampler=dist_sampler,
+            shuffle=False
```
Comment (rohitgr7) [inline tests/trainer/test_dataloaders.py:759]:
```suggestion
            shuffle=True
```

File: tests/trainer/test_dataloaders.py [hunk 4]
```diff
@@ -746,6 +746,39 @@ def test_dataloader_distributed_sampler(tmpdir):
...
+
+    def train_dataloader(self):
+        dataloader = super().train_dataloader()
+        dist_sampler = DistributedSampler(dataloader.dataset, shuffle=False)
+        return DataLoader(
+            dataloader.dataset,
+            batch_size=self.batch_size,
+            drop_last=False,
+            sampler=dist_sampler,
+            shuffle=True
```
Comment (rohitgr7) [inline tests/trainer/test_dataloaders.py:759]:
```suggestion
            shuffle=False
```

File: tests/trainer/test_dataloaders.py [hunk 5]
```diff
@@ -746,6 +746,39 @@ def test_dataloader_distributed_sampler(tmpdir):
     trainer.test(ckpt_path=None)
 
 
+class ModelWithDataLoaderDistributedSampler(EvalModelTemplate):
+
+    def train_dataloader(self):
+        dataloader = super().train_dataloader()
+        dist_sampler = DistributedSampler(dataloader.dataset, shuffle=False)
```
Comment (rohitgr7) [inline tests/trainer/test_dataloaders.py:753]:
```suggestion
        dist_sampler = DistributedSampler(dataloader.dataset, shuffle=True)
```

**Review by rohitgr7 — COMMENTED** [review]
just trying.

**Comment by rohitgr7** [comment]
I suggest while adding a sampler we should explicitlly set `shuffle=False`[here](https://github.com/PyTorchLightning/pytorch-lightning/blob/3abfec896212ea85e45d6ac3ccb323ef242d16de/pytorch_lightning/trainer/data_loading.py#L138) to avoid the case if the dataloader already has `shuffle=True`.

**Review by rohitgr7 — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

**Comment by SeanNaren** [comment]
@rohitgr7 think we should include that in this PR or handle this separately?

**Comment by rohitgr7** [comment]
@SeanNaren either way is fine or let's just do it separately and also here in `_get_distributed_sampler(self, dataloader, train):` this `train` argument should be renamed to `shuffle` for better readability of the code.

**Review by rohitgr7 — CHANGES_REQUESTED** [review]
please wait

**Review by rohitgr7 — APPROVED** [review]
